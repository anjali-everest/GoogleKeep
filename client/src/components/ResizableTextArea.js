import React, { useState } from "react";
import PropTypes from "prop-types";

const ResizableTextarea = (props) => {
  const [rows, setRows] = useState(1);
  const [value, setValue] = useState("");
  const minRows = 1,
    maxRows = 50;

  const handleChange = (event) => {
    const textareaLineHeight = 12;
    const previousRows = event.target.rows;
    event.target.rows = minRows;
    const currentRows = Math.floor(
      event.target.scrollHeight / textareaLineHeight
    );
    if (currentRows === previousRows) {
      event.target.rows = currentRows;
    }
    if (currentRows >= maxRows) {
      event.target.rows = maxRows;
      event.target.scrollTop = event.target.scrollHeight;
    }
    setRows(currentRows < maxRows ? currentRows : maxRows);
    setValue(event.target.value);
    props.onChange(event);
  };

  return (
    <textarea
      rows={rows}
      value={value}
      placeholder={props.placeholder}
      className={props.id}
      id={props.id}
      onChange={handleChange}
      onClick={props.onClick}
    />
  );
};

ResizableTextarea.propTypes = {
  id: PropTypes.string,
  onChange: PropTypes.func,
  onClick: PropTypes.func,
  placeholder: PropTypes.string,
};

export default ResizableTextarea;
