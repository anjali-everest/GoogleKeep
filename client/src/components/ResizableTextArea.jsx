import React from "react";

class ResizableTextarea extends React.PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      value: "",
      rows: 1,
      minRows: 1,
      maxRows: 50,
    };
  }

  handleChange = (event) => {
    const textareaLineHeight = 12;
    const { minRows, maxRows } = this.state;
    const previousRows = event.target.rows;
    event.target.rows = minRows; // reset number of rows in textarea
    const currentRows = ~~(event.target.scrollHeight / textareaLineHeight);
    if (currentRows === previousRows) {
      event.target.rows = currentRows;
    }
    if (currentRows >= maxRows) {
      event.target.rows = maxRows;
      event.target.scrollTop = event.target.scrollHeight;
    }
    this.setState({
      value: event.target.value,
      rows: currentRows < maxRows ? currentRows : maxRows,
    });
    this.props.onChange(event);
  };

  render() {
    return (
      <textarea
        rows={this.state.rows}
        value={this.state.value}
        placeholder={this.props.placeholder}
        className={"textarea"}
        onChange={this.handleChange}
        onClick={this.props.onClick}
      />
    );
  }
}

export default ResizableTextarea;
