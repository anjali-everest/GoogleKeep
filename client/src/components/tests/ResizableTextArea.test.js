import React from "react";
import { shallow } from "enzyme";
import ResizableTextArea from "../ResizableTextArea";

describe("ResizableTextArea tests", () => {
  it("Should render without crashing", () => {
    shallow(<ResizableTextArea />);
  });
  it("should match the ResizableTextArea snapshot", function () {
    const resizableTextArea = shallow(<ResizableTextArea />);
    expect(resizableTextArea).toMatchSnapshot();
  });
});
