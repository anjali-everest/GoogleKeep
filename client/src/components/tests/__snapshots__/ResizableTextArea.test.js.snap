// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResizableTextArea tests should match the ResizableTextArea snapshot 1`] = `ShallowWrapper {}`;
